### PR TITLE
Module 1

### PR DESCRIPTION
**Pull request by JoeBawn**
Please make sure you are raising this PR against your own repository and not the original. If it says https://github.com/CorndelWithSoftwire in the search bar right now, you are in the wrong place!

Please also don't forget to submit the PR's url to Aptem afterwards (by clicking on the corresponding exercise component, pasting the url in the textbox that appears and then pressing the finish button).

### CODE REVIEW

File: todo_app/templates/index.html [hunk 1]
```diff
@@ -3,16 +3,52 @@
...
+      </form>
+      <table class="table">
+        <ul class="list-group mb-4">
+          <thead>
+            <tr>
+              <th scope="col">Task Lists</th>
+              {% if request.values.get('sort') == None %}
+              <th  scope="col"><a href="/?sort=2" class="table">Status</a></th>
+              {% elif request.values.get('sort') == '2' %}
+              <th scope="col"><a href="/?sort=1" class="table">Status ↕</a></th>
```
Comment (hugh-emerson) [inline todo_app/templates/index.html:28]: Would expect the arrow ↕ when it's not sorted and up/down arrows for the other two cases to show that it's sorted ascending/descending

File: todo_app/app.py [hunk 2]
```diff
@@ -1,15 +1,47 @@
...
+
+@app.route('/remove_item', methods=['POST'])
+def remove_existing_item():
+    remove_id = request.form.get('remove_id')
+    items = remove_item(remove_id)
+    return redirect(request.headers.get('Referer'))
+
+@app.route('/toggle_status', methods=['POST'])
+def toggle_status():
+    toggle_item = get_item(request.form.get('toggle_item_id'))
```
Comment (hugh-emerson) [inline todo_app/app.py:36]: Why is the form input called "toggle_item_id" instead of just "item_id"?

File: todo_app/app.py [hunk 3]
```diff
@@ -1,15 +1,47 @@
...
+        items = sorted(items, key=itemgetter('status'))
+    elif request.values.get('sort') == '2':
+        items = sorted(items, key=itemgetter('status'), reverse=True)
+    else:
+        items = sorted(items, key=itemgetter('id'))
+    return render_template("index.html",items=items)
 
+@app.route('/', methods=['POST'])
+def add_list():
+    newitemtitle = request.form.get('title')
```
Comment (hugh-emerson) [inline todo_app/app.py:24]: The Python naming convention is snake_case so this should be `new_item_title`

File: todo_app/app.py [hunk 4]
```diff
@@ -1,15 +1,47 @@
...
 @app.route('/')
 def index():
-    return 'Hello World!'
+    items = get_items()
+    if request.values.get('sort') == '1':
+        items = sorted(items, key=itemgetter('status'))
+    elif request.values.get('sort') == '2':
+        items = sorted(items, key=itemgetter('status'), reverse=True)
+    else:
+        items = sorted(items, key=itemgetter('id'))
```
Comment (hugh-emerson) [inline todo_app/app.py:19]: Why have this else block?

**Review by hugh-emerson — APPROVED** [review]
I've left some comments but it does all work and you can merge whenever you're ready 👍

File: todo_app/templates/index.html [hunk 5]
```diff
@@ -3,16 +3,52 @@
...
+              </tr>
+          {% for item in items %}
+          </thead>
+          <tbody>
+            <tr>
+                    <td><li class="list-group-item">{{item.title}}</td>
+                    <td><li class="list-group-item">{{ item.status }}</td>
+                    <td><form id="toggle_item_{{ item.id }}" action="/toggle_status" method="POST">
+                      <input type="hidden" name="toggle_item_id" value="{{ item.id }}">
+                      <a onclick="document.getElementById('remove_item_{{ item.id }}').submit()"><button ng-click="remove_item()" class="btn btn-outline-success btn-block" style="margin-top:5px">Completed</button></a>
```
[the remark this answers is not in the document]
Comment (JoeBawn) [reply todo_app/templates/index.html]: Removed <a> tag and javascript as suggested.

File: todo_app/data/session_items.py [hunk 6]
```diff
@@ -67,3 +67,20 @@ def save_item(item):
...
     return item
+
+def remove_item(id):
+    """
+    Removes an existing item in the session. If no existing item matches the ID of the specified item, nothing is removed.
+    Args:
+        item: The item to save.
+    """
+    existing_items = get_items()
+    for i in range(len(existing_items)):
```
[the remark this answers is not in the document]
Comment (JoeBawn) [reply todo_app/data/session_items.py]: Understood and have amended remove function:

existing_items = get_items()
    for i in existing_items:
        if i['id'] == int(id):
            existing_items.remove(i)
            break

    session['items'] = existing_items

File: todo_app/data/session_items.py [hunk 7]
```diff
@@ -67,3 +67,20 @@ def save_item(item):
...
+        if existing_items[i]['id'] == int(id):
+            del existing_items[i]
+            break
+
+    session['items'] = existing_items
+   
+
+    return
```
[the remark this answers is not in the document]
Comment (JoeBawn) [reply todo_app/data/session_items.py]: return removed as suggested

File: todo_app/templates/index.html (hunk 1 above)
Comment (JoeBawn) [reply todo_app/templates/index.html]: Amended arrows to reflect correct sort status

File: todo_app/app.py (hunk 2 above)
Comment (JoeBawn) [reply todo_app/app.py]: form input updated to item_id

File: todo_app/app.py (hunk 3 above)
Comment (JoeBawn) [reply todo_app/app.py]: updated to follow python naming convention

File: todo_app/app.py (hunk 4 above)
Comment (JoeBawn) [reply todo_app/app.py]: removed else block.

**Comment by JoeBawn** [comment]
Thanks for the feedback Hugh!

I have gone through and amended with the last two commits and will merge the pull request.

**Pull request merged**